### PR TITLE
Sparse ELLPACK

### PR DESCRIPTION
**Pull request by atait**
**Motivation and context:**
Performance improvement to sparse matrix vector multiplications. See #186 for background and initial benchmarking. This PR completes the TODO list mentioned in that issue.

Depends on #187 being merged first.

**How long should this take to review?**
- Moderate/lengthy. There are a lot of lines added, but it is tested pretty well. I rewrote history in attempts to make it easier to review.

**Where should a reviewer start?**
Run the tests. 

Commits reflect the logical progress of the different features if you want to follow along,
1. "working serial ELLPACK" is the data storage format and reference implementation
2. "working tree-based reduction" is the most basic performant algorithm. It is now the default implementation, but does not work for more than 1024 fan-in per neuron
3. "two kernel big ELLPACK" has the algorithm that works up to 1M fan-in per neuron
4. "spmv_prog style" is the new planner interface, selecting algorithms, determining device storage formats, etc.

**Questions for devs**
- should `clra_spmv.py` be its own file?
- does the `spmv_prog` class make sense? I tried to follow the patterns used in `gemv_prog` and the gemv `_impl`s
- There are other ELLPACK implementations "serial" and "nonlocal" that could be cut. The regular user will only ever use the best one.

### CODE REVIEW

**Comment by hunse** [comment]
This seems to do a lot better than the CSR format, at least for the Watts-Strogatz benchmark and the tests, so I think it's a great addition!

For the test, if I increase the size to 5000, then I start getting a number of errors in `test_sparse`. I've added this case, so hopefully you're able to reproduce it. Would you have time to look into this?

One limitation to this implementation that I can think of is if you've got a sparse matrix with one particularly non-sparse row (for example, a diagonal matrix where every element in the first row also has a non-zero value). In this case, you'd end up using more memory and computation than just storing the matrix densely. Would it be possible to fall back on the CSR implementation in cases like these? We could have some sort of heuristic that compares e.g. `rows * ellwidth` to `nnz`, and if it's many times bigger, then use CSR instead. The best would probably be to base this heuristic on some empirical trials (e.g. increase the length of one row until CSR does better than ELLPACK), but I wouldn't go so far as to do autotuning (i.e. we'd just pick the heuristic in advance for a fairly standard GPU, not try to tune it to individual machines). I'm not too worried about having optimal performance in this case, my main concern is that we support all kinds of sparse matrices (i.e. without running out of memory if a matrix has one long row).

File: nengo_ocl/clra_gemv.py [hunk 1]
```diff
@@ -716,7 +718,7 @@ def many_dots_impl(p, items):  # noqa: C901
 
     segment_size = min(max_y_len, MAX_SEGMENT_SIZE)
     dot_block_size = min(
-        max(max_n_dots, 1), int(p.queue.device.max_work_group_size / segment_size)
+        max(max_n_dots, 1), int(p.queue.device.max_work_group_size / 2 / segment_size)
```
Comment (hunse) [inline nengo_ocl/clra_gemv.py:721]: Why this change?

File: nengo_ocl/clra_gemv.py [hunk 2]
```diff
@@ -1238,43 +1240,376 @@ def choose_plans(self):
...
-            raise NotImplementedError(
-                "OCL SparseDot only supports matrix-vector currently, not matrix-matrix"
+    @staticmethod
+    def scipy2elldata(scipy_mat, force_nonempty=True):
+        """ Works with scipy sparse or numpy dense argument """
+        scipy_mat = scipy_mat.copy()
+        try:
+            lilmat = scipy_mat.tolil()
+        except AttributeError:
+            lilmat = scipy_sparse.lil_matrix(scipy_mat)
```
Comment (hunse) [inline nengo_ocl/clra_gemv.py:1333]: Is there a reason to not just use this line for all cases? What's the reasoning behind trying `scipy_mat.tolil()` first?

File: nengo_ocl/clra_gemv.py [hunk 3]
```diff
@@ -1238,43 +1240,376 @@ def choose_plans(self):
...
+
+        shape = scipy_mat.shape
+        rowlens = np.array([len(rowdata) for rowdata in lilmat.rows])
+        ellwidth = max(rowlens)
+        ind_dtype = (
+            np.int32 if len(rowlens) < 2 ** 32 and ellwidth < 2 ** 32 else np.int64
```
Comment (hunse) [inline nengo_ocl/clra_gemv.py:1339]: Since that'd be a huge matrix, I'm guessing this hasn't been tested (also, it should technically be `2**31`, since that's the max value for signed `int32`). I think we should just hard-code this as `int32`, since I don't see us hitting that limit any time soon (someone would need to have more than 2 billion neurons in an ensemble), and probably there would be further considerations if we're moving into that size of matrix.

File: nengo_ocl/clra_gemv.py [hunk 4]
```diff
@@ -1316,16 +1662,380 @@ def plan_sparse_dot_inc(queue, A_indices, A_indptr, A_data, X, Y, inc=False, tag
...
+    _fn = cl.Program(queue.context, text).build().ellpack_inc
+    _fn.set_args(*full_args)
+
+    plan = Plan(queue, _fn, gsize, lsize=lsize, name="cl_ellpack", tag=tag)
+    plan.full_args = full_args  # prevent garbage-collection
+
+    return plan
+
+
+def spmv_ellpacknonlocal_impl(queue, A_columns, A_entries, X, Y, inc=False, tag=None):
```
Comment (hunse) [inline nengo_ocl/clra_gemv.py:1786]: At some point, we'll want to remove the non-performant ones. What do you think about removing them now, but doing it as a second commit, so that anybody wanting to see the other alternatives can still look at your original commit?

File: nengo_ocl/tests/test_clra_gemv.py [hunk 5]
```diff
@@ -10,13 +10,18 @@
 
 from nengo_ocl.clra_gemv import (
     plan_block_gemv,
+    plan_csr,
+    plan_ellpack,
+    plan_ellpack_nonlocal,
+    plan_ellpack_serial,
+    plan_ellpack_tree,
+    plan_ellpack_twostep,
```
Comment (hunse) [inline nengo_ocl/tests/test_clra_gemv.py:18]: NTS: remove the unused ones

File: nengo_ocl/tests/test_clra_gemv.py [hunk 6]
```diff
@@ -34,6 +39,18 @@ def pytest_generate_tests(metafunc):
...
         )
+    if "sparse_planner" in metafunc.fixturenames:
+        metafunc.parametrize(
+            "sparse_planner",
+            [
+                plan_ellpack,
+                # plan_ellpack_twostep,
+                # plan_ellpack_nonlocal,
+                # plan_ellpack_tree,
+                # plan_ellpack_serial,
```
Comment (hunse) [inline nengo_ocl/tests/test_clra_gemv.py:50]: NTS: remove these comments

**Review by hunse — CHANGES_REQUESTED** [review]

**Comment by atait** [comment]
I will take a look at the error this week. Thanks for the comments.

You are right that Watts Strogatz is sort of a best case scenario for ELLPACK. A heuristic based on `rows`, `ellwidth`, and `nnz` would work. CSR is always going to be smaller footprint. The longest-running CSR thread occurs in the row with most non-zeros, and it is doing serial reduction. That means ELLPACK is always going to be faster, even if it is doing lots of zero operations, because its threads don't diverge. For `ellwidth>1024`, there needs to be that two-step reduction, extra kernel overhead, more memory movement – I think speed comparisons with CSR would have to be empirical in that case.

It would be nice if the fallback condition took into account available device memory and had an override configuration option somewhere.
```python
footprint_ratio = 8 * ellwidth * rows / ctx.devices[0].global_mem_size
use_ellpack = (
    (
        footprint_ratio < 0.5  # 0.5 is some threshold value
        or (footprint_ratio < 0.9 and ellwidth < 1024)  # more tolerant if using single-step
        or 'ELL' in os.environ.get('nengo_ocl_spmv_algo', '')
    )
    and not 'CSR' in os.environ.get('nengo_ocl_spmv_algo', '')
)
```
There are some more advanced representations to balance speed and memory efficiency, described well in https://arxiv.org/pdf/1307.6209.pdf. It doesn't require autotuning in the empirical sense, just more algorithm writing. Here's a picture of a pretty simple one that decomposes into sum of ELLPACK and CSR. It is described here https://medium.com/analytics-vidhya/sparse-matrix-vector-multiplication-with-cuda-42d191878e8f, although I can't find the general-case decomposition algorithm in that person's code. I'm sure it could be found in the literature, if you think it is worth looking into. For my purposes, simple fallback with override is perfect.

IMG_3668

File: nengo_ocl/clra_gemv.py (hunk 2 above)
Comment (atait) [reply nengo_ocl/clra_gemv.py:1333]: Originally this was an attempt to avoid a copy and keep the underlying data. The first one returns the same object. I just looked at scipy, and the second one would keep the same data even though it is a different object. Anyways, it's moot because you don't want to maintain the link sometimes and not other times, depending on input.

File: nengo_ocl/clra_gemv.py (hunk 1 above)
Comment (atait) [reply nengo_ocl/clra_gemv.py:721]: Oops. I was just testing it out and forgot to revert.

File: nengo_ocl/clra_gemv.py (hunk 3 above)
Comment (atait) [reply nengo_ocl/clra_gemv.py:1339]: Good point!

**Comment by atait** [comment]
Ok I think those are all the changes. That error was caused by one number. I must have made a mistake with that line while revising the history. You may want to check the heuristic for default values or other conditions you can think of. 

I removed the serial and global reduction algorithms, but I kept `plan_ellpack_tree` and `plan_ellpack_twostep` in the test file (commented) and environment variable options. While `Simulator` doesn't interact with those directly, they're still in the library, so it's just 2 lines of code to allow forcing one or the other.

Thanks for the detailed review. I think it's solid now. Out of curiosity, I searched the TF, Pytorch, NEST, and CoreNEURON repos (~50k issues in sum) and only found one mention of "ELLPACK" https://github.com/pytorch/pytorch/issues/6171 

> Yes, this makes sense (indeed, with the new `torch.layout` API bits, we have planned for it), but no, it's not so easy to support. It basically entails adding a new backend to PyTorch, and there are a lot of moving bits to making this happen, especially because we are rewriting how all of this works internally (C10).

Too many moving bits I guess.

**Comment by hunse** [comment]
The fixes look good. I did find, though that the conversion to the LIL matrix format was causing problems when I ran the Watts-Strogatz example (before, you were copying the matrix before converting to LIL, and you got rid of that, which seemed to cause problems). It seemed unnecessary to convert to LIL, so I rewrote that code to make the ELLPACK format straight from the CSR representation. I also added a `HostSparseMatrix` class that caches the CSR representation so that we're not needlessly re-converting to CSR if the input matrix is in a different format.

If these changes look good to you @atait, I'll merge.

**Review by hunse — APPROVED** [review]

**Comment by atait** [comment]
The caching CSR approach makes sense. The user should be the one to deal with sparsity-changing data structures when building the matrix.

The question is behavior with learning rules. In regular nengo, I think plastic `BsrDotInc` is supported. https://github.com/nengo/nengo/pull/1629. However, I don't think plasticity is supportable here if the base data is anything other than CSR. If `self.matrix` is CSR already, it is possible for base `data` changes to affect the `self._csr`. Otherwise, scipy makes a copy, and learning updates are not possible. The behavior of always breaking ties to base data could be done with this
```python
    self._csr = scipy_sparse.csr_matrix(self.matrix, copy=True) 
```
Alternatively, you could avoid a copy (only if it is CSR), keep the linkage, but error for attempted base data modification like this
```
@property
def csr(self):
    if self._csr is None:
        self._csr = scipy_sparse.csr_matrix(self.matrix, copy=False)
        self._csr.indices.flags.writable = False
        self._csr.indptr.flags.writable = False
        self._csr.indptr.data.flags.writable = False
```

I'm not sure what a use case for plastic sparse connections would be, so you can probably make a better call on which one (if either) to do.

Everything else looks great to me.

**Comment by hunse** [comment]
In Nengo, `BsrDotInc` is an operator that we create ourselves to combine multiple `DotInc` operators into one. We don't currently allow learning on Sparse transforms in core Nengo, so as long as things work here, I think that's all we have to worry about.

**Comment by hunse** [comment]
When I was getting this ready for merge, I found another problem. `plan_ellpack` wasn't working for matrices with `ellwidth > max_work_group_size`. The reason for this is that when determining the number of workgroups per row, the value got rounded down instead of getting rounded up. I fixed this in the first fixup commit, as well as giving some of the variables a bit more general names (referring to rows/columns rather than neurons/synapses).

I also realized that the `round_up_power_of_2` wasn't necessary when determining the number of workgroups per row, and could result in some workgroups not doing much or anything. (For example, if the workgroup size is 1024 and the ellwidth is 2049, the algorithm would use 4 workgroups per row, with the first two doing most of the computation, the third processing one element, and the fourth doing nothing.) So I changed this to use just the number of workgroups required, which necessitated some minor changes to `ellpack_gblreduce`. I added a large test case (commented out) that demonstrates the speed improvement after this change.

Finally, `plan_ellpack_twostep.assign_plan_characteristics` was never getting called, so I fixed that.

**Comment by atait** [comment]
> plan_ellpack wasn't working for matrices with ellwidth > max_work_group_size

Sorry for that bug. I see it now and guess my tests just missed those `ellwidth`s.

> I also realized that the round_up_power_of_2 wasn't necessary when determining the number of workgroups per row

Yes, good catch. Power of 2 is only necessary for the global work group.
